oak/server/rust/oak_runtime/src/io.rs

```diff
@@ -0,0 +1,110 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{runtime::RuntimeProxy, NodeMessage};
+use log::error;
+use oak_abi::{ChannelReadStatus, Handle, OakStatus};
+
+/// A trait for objects that can be decoded from bytes + handles.
+pub trait Decodable: Sized {
+    fn decode(message: &NodeMessage) -> Result<Self, OakStatus>;
+}
+
+impl<T: prost::Message + std::default::Default> Decodable for T {
+    fn decode(message: &NodeMessage) -> Result<Self, OakStatus> {
+        if !message.handles.is_empty() {
+            error!("Non-empty handles field");
+            return Err(OakStatus::ErrInternal);
+        }
+        let value = T::decode(message.data.as_slice()).map_err(|error| {
+            error!("Couldn't decode Protobuf message: {}", error);
+            OakStatus::ErrInternal
+        })?;
+        Ok(value)
+    }
+}
```

Comment:
Just checking, is this actually used for anything yet? If not, I think we may leave it out for now (blanket impl are a bit weird sometimes, they confuse the type checker in some cases).

Reply:
It is used in:
```rust
pub struct Invocation {
    receiver: Receiver<GrpcRequest>,
    sender: Sender<GrpcResponse>,
}
```
for decoding `GrpcRequest` and `GrpcResponse`.